Fix gameplay skipping forward during resume operation

peppy opened the pull request:
Closes https://github.com/ppy/osu/issues/19975.

I'm not precisely sure *where* this regressed in the changes, but it did. Rather than trying to figure out where the issue lies, I've just explicitly added this flow so we are aware of what is going on where and when.

Inline comment should explain this quite well.

peppy committed:
Add test coverage of resuming after pause not skipping forward in time

peppy committed:
Fix gameplay skipping forward during resume operation

frenzibyte commented:
I've took the time to debug through why this is occurring with the refactor and it turns out this is how things are supposed to work w.r.t. pause and resume:
 1. `GameplayClock.Stop()` is called due to user pause, setting `IsPaused.Value = true` and transforming the frequency to zero
 2. While the frequency is being adjusted, the clocks stop getting processed on `Update`
 3. `GameplayClock.Start()` is called due to user resume, seeking the track to the last processed frame time, which was at the moment `Stop()` was called.

This led to the realisation that `FramedBeatmapClock` actually always calls `ProcessFrame` regardless of the paused state, unlike the previous `GameplayClockContainer` version which processes the clocks only when in running state:

<img width="448" alt="CleanShot 2022-08-29 at 15 07 36@2x" src="https://user-images.githubusercontent.com/22781491/187197547-676b7fd6-5af0-44ac-8160-3da053fc7705.png">

So while the solution here is valid, the new behaviour with `FramedBeatmapClock` makes me feel slightly uneasy about it potentially regressing other scenarios, but I can't think of any in my head, so we'll see how things go.

frenzibyte commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -111,11 +109,22 @@ public void Start()
+        /// </summary>
+        protected virtual void PrepareStart()
+        {
+            // Seeking the decoupled clock to its current time ensures that its source clock will be seeked to the same time
+            // This accounts for the clock source potentially taking time to enter a completely stopped state
+            Seek(GameplayClock.CurrentTime);
```
As per my comment above, this is now pretty much pointless since the clock is always processed regardless of pause state.

peppy replied:
I've removed this for now.

frenzibyte reviewed (commented):
Also there seems to be leftover/debugging changes pushed in this PR.

peppy committed:
Remove debug changes

peppy commented:
I figured that's how things were, although for whatever reason couldn't find the code in your screenshot.

peppy committed:
Remove probably unnecessary `Seek` on start

peppy committed:
Fix failing large offset test

If we are going to continue to let the underlying clock process frames,
there needs to be a bit of lenience to allow the backwards seek on
resume (to play back over the freq ramp period).

The test is meant to be ensuring we don't skip the full offset amount,
so div10 seems pretty safe.

peppy committed:
Merge branch 'master' into fix-resume-skip-forward

smoogipoo commented on osu.Game/Screens/Play/MasterGameplayClockContainer.cs:
```diff
@@ -86,6 +97,8 @@ private double findEarliestStartTime()
 
         protected override void StopGameplayClock()
         {
+            actualStopTime = GameplayClock.CurrentTime;
```
Why does this work? I've read the xmldoc and I can't understand how this being here and _not_ inside the `OnComplete()` callback below translates to the clock seeking to the correct time at unpause? Logically I would think it think it would seek to just under 200ms or so prior to the point where gameplay is fully stopped.

peppy replied:
In `FrameStabilityClock`, as soon as it sees `IsPaused` it stops processing gameplay. So yes, you are correct in thinking that this will seek 200ms prior, but from the gameplay perspective this is where it wants to be.

I intend on addressing this in the refactor I have planned (ie. optimally no backward seek should be required. it should always go forward and everything should see that ongoing time, including `FrameStabilityClock`).

smoogipoo replied:
Today I learnt that gameplay is paused immediately.

peppy replied:
Yeah, as I mentioned somewhere recently, it's not how I want it to be, but I believe it was done to avoid entering a `Failed` state after `Paused`. If you remember that debacle.

frenzibyte replied:
Perhaps renaming `StopGameplayClock` to `PauseGameplayClock` might indicate it a bit further, as I would've also thought `StopGameplayClock` to be called when the clock should actually just be stopped.
